### PR TITLE
spirv-fuzz: Enable some passes with high probability

### LINKED ISSUE #3764: spirv-fuzz: Enable certain passes to run with a higher probability than the others
Currently, all the fuzzer passes are run with the same probability.
However, we may consider some of them to be more important than others
and want them to have a higher probability of running.

It would therefore be useful to introduce in some way the concept of
prioritary fuzzer passes.

### PR DESCRIPTION
**Pull request by afd**
Fixes #3764.

### CODE REVIEW

File: source/fuzz/fuzzer.h [hunk 1]
```diff
@@ -87,17 +87,26 @@ class Fuzzer {
...
-  void MaybeAddRepeatedPass(RepeatedPassInstances* pass_instances,
+  void MaybeAddRepeatedPass(uint32_t percentage_chance_of_adding_pass,
+                            RepeatedPassInstances* pass_instances,
                             Args&&... extra_args);
 
+  // The same as the above, with |percentage_chance_of_adding_pass| == 50%.
+  template <typename FuzzerPassT, typename... Args>
+  void MaybeAddRepeatedPass(RepeatedPassInstances* pass_instances,
+                            Args&&... extra_args) {
+    MaybeAddRepeatedPass<FuzzerPassT>(50, pass_instances, extra_args...);
```
Comment (Vasniktel) [inline source/fuzz/fuzzer.h:105]: It would probably be better to use `std::forward<Args>(extra_args)...` here.

File: source/fuzz/fuzzer.cpp [hunk 2]
```diff
@@ -217,12 +219,18 @@ Fuzzer::FuzzerResult Fuzzer::Run() {
                                                       initial_facts_);
 
   RepeatedPassInstances pass_instances{};
+
+  // The following passes are likely to be very useful: many other passes
+  // introduce synonyms, irrelevant ids and constants that these passes can work
+  // with.  We thus enable them with high probability.
+  MaybeAddRepeatedPass<FuzzerPassObfuscateConstants>(90, &pass_instances);
+  MaybeAddRepeatedPass<FuzzerPassApplyIdSynonyms>(90, &pass_instances);
+  MaybeAddRepeatedPass<FuzzerPassReplaceIrrelevantIds>(90, &pass_instances);
```
Comment (Vasniktel) [inline source/fuzz/fuzzer.cpp:228]: I wonder if it's better to put these into the loop below. The motivation is that some of the transformations applied there might introduce instructions that we can later obfuscate, create synonyms to and replace with irrelevant ids.

In particular, I believe we don't have any irrelevant ids initially.

**Review by Vasniktel — COMMENTED** [review]
Looks good. I left a small question in the review comments.

File: source/fuzz/fuzzer.cpp (hunk 2 above)
Comment (afd) [reply source/fuzz/fuzzer.cpp:228]: MaybeAddRepeatedPass just makes the pass exist in pass_instances.

There is then, later, a loop that selects random passes from the sequence of pass indices, so it's not the case that these passes will be run first (necessarily).

See what I mean?

Comment (Vasniktel) [reply source/fuzz/fuzzer.cpp:228]: Yes, you are right indeed. Sorry for the confusion.

**Review by Vasniktel — APPROVED** [review]

File: source/fuzz/fuzzer.h (hunk 1 above)
Comment (afd) [reply source/fuzz/fuzzer.h]: Done.

**Review by Vasniktel — APPROVED** [review]

**Pull request merged**